ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala

```diff
@@ -747,6 +748,42 @@ abstract class AbstractWebsocketServiceIntegrationTest
     go(createCount)
   }
 
+  /** Updates the ACS retrieved with [[readAcsN]] until the given offset
+    */
+  private[this] def updateAcs(
+      acs: Vector[(String, JsValue)],
+      expectLivenessMarker: Boolean,
+      until: domain.Offset,
+  ): Consume.FCC[JsValue, Map[String, JsValue]] = {
+    val dummyLivenessMarker =
+      JsObject("events" -> JsArray.empty, "offset" -> JsNumber(0)).asInstanceOf[JsValue]
+    val dslSyntax = Consume.syntax[JsValue]
+    import dslSyntax._
+    import domain.Offset.ordering
+    import scalaz.syntax.order._
+    def go(
+        acs: Map[String, JsValue],
+        latest: Option[domain.Offset],
+    ): Consume.FCC[JsValue, Map[String, JsValue]] =
+      if (latest.fold(false)(_ >= until)) {
```

Comment:
I think I’m confused. If `latest` is actually greater than `until`, we don’t want to include this do we? It looks to me like we might be reading the heartbeat again which adds more 5s delays:
1. line 774 reads something with offset `until`
2. We get here, we read again in line 770, at this point we are reading the heartbeat and things work only by accident because it is empty.

Am I missing something here?

Reply:
Mmmh, I think you're right. I did this to not miss the last object, but probably using a dummy create as you suggested initially is easier, let me try that.

Reply:
I forgot what the case actually was where we can actually miss something. If we read a non-heartbeat message and the offset is >= until we can stop. The only case where we might want to read a heartbeat is if `latest` is `None` but we already do that in the second case.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/9847/commits/ab2535a7f091464d915222749bebac2fd0b091d6

I switched from using offsets to asking for the expected number of events. I also renamed `ledgerEnd` to `acsEnd`, which is more accurate.